### PR TITLE
Reuse socket

### LINKED ISSUE #87: Re-use socket for journald messages
Hello @lucab 

currently the `logging::journald_send` always creates a fresh unbound socket to send the journald datagrams which in my understanding allocates and closes a fresh file descriptor for every `journald_send` call.  File descriptors aren't too efficient in the kernel, so I think we should avoiding creating them frequently.

I'd like to add an API which lets me create the socket once and reuse it for all subsequent messages in my logger implementation.   Specifically, I'd like to add

- a `struct JournalWriteClient` which holds the socket,
- `JournalWriteClient::connect() -> JournalWriteClient` which creates and binds the socket,
- `JournalWriteClient::send` and `JournalWriteClient::print` for `journal_send` and `journal_print`

We can then either remove the old `journal_send` and `journal_print`, or alternatively re-implement them on to of `JournalWriteClient`.

I'm not sure about the `JournalWriteClient` name; I also considered `JournalWriter`, but this seems to suggest that it implements `std::io::Write`, or `JournalClient` but that might suggest that it can also read from the journal.

What do you think?

### PR DESCRIPTION
**Pull request by lunaryorn**
Closes #87

### CODE REVIEW

**Comment by lunaryorn** [comment]
@lucab Is this what you had in mind?

File: src/logging.rs [hunk 1]
```diff
@@ -159,7 +176,7 @@ where
     //
     // Maximum data size is system dependent, thus this always tries the fast path and
     // falls back to the slow path if the former fails with `EMSGSIZE`.
-    let fast_res = sock.send_to(&data, SD_JOURNAL_SOCK_PATH);
+    let fast_res = sock.send(&data);
```
Comment (lucab) [inline src/logging.rs:179]: I seem to remember there was a reason behind keeping this socket unconnected and using a destination address, but I can't find the details right now.
I'll get back when I managed to re-surface those, but in the meanwhile please consider sticking to the previous unconnected approach.

Comment (lunaryorn) [reply src/logging.rs:179]: Sure I can.

Comment (lunaryorn) [reply src/logging.rs]: Done

**Review by lucab — APPROVED** [review]

**Comment by lucab** [comment]
LGTM. I went back through my github history, and the reason for using an unconnected socket client-side is to cope with journald restarts, see https://github.com/coreos/go-systemd/pull/279. The native systemd C library does the same, see https://github.com/coreos/go-systemd/pull/218#issuecomment-274490875.

If we start exposing a way for consumers to provide a socket, we'll have to keep that in mind and check that it is a proper unbound one.

**Comment by lunaryorn** [comment]
Makes sense.

I've merged this pull request; can we make a new release?

**Pull request merged**